pkg/monitor/datapath_trace.go

```diff
@@ -60,13 +60,51 @@ type TraceNotifyV0 struct {
 	// data
 }
 
+// decodeTraceNotifyVersion0 decodes the trace notify message in 'data' into
+// the struct. This function needs to be kept in sync with the struct above.
+func (tn *TraceNotifyV0) decodeTraceNotifyVersion0(data []byte) error {
+	// This eliminates the bounds check in the accesses to `data` below.
+	if len(data) < traceNotifyV0Len {
+		return errors.New("unexpected length of data to decode into TraceNotifyV0")
+	}
+
+	tn.Type = data[0]
+	tn.ObsPoint = data[1]
+	tn.Source = byteorder.Native.Uint16(data[2:4])
+	tn.Hash = byteorder.Native.Uint32(data[4:8])
+	tn.OrigLen = byteorder.Native.Uint32(data[8:12])
+	tn.CapLen = byteorder.Native.Uint16(data[12:14])
+	tn.Version = byteorder.Native.Uint16(data[14:16])
+	tn.SrcLabel = identity.NumericIdentity(byteorder.Native.Uint32(data[16:20]))
+	tn.DstLabel = identity.NumericIdentity(byteorder.Native.Uint32(data[20:24]))
+	tn.DstID = byteorder.Native.Uint16(data[24:26])
+	tn.Reason = data[26]
+	tn.Flags = data[27]
+	tn.Ifindex = byteorder.Native.Uint32(data[28:32])
+
+	return nil
+}
+
 // TraceNotifyV1 is the version 1 message format.
```

Comment:
probably a good idea to have the same kind of comment as for v0 about keeping the `struct`/`func` in sync

```suggestion
// TraceNotifyV1 is the version 1 message format. This struct needs to be kept
// in sync with the decodeTraceNotifyVersion1 func.
```